config/initializers/doorkeeper.rb

```diff
@@ -38,10 +38,19 @@
   # If you want to disable expiration, set this to nil.
   access_token_expires_in nil
 
-  # Assign a custom TTL for implicit grants.
-  # custom_access_token_expires_in do |oauth_client|
-  #   oauth_client.application.additional_settings.implicit_oauth_expiration
-  # end
+  # context.grant_type to compare with Doorkeeper::OAUTH grant type constants
+  # context.client for client (Doorkeeper::Application)
+  # context.scopes for scopes
+  custom_access_token_expires_in do |context|
+    # If the client is confidential (all clients pre 4.3), then we don't want to
+    # expire access tokens. Applications created by users are also considered
+    # confidential.
+    if context.client.confidential?
+      nil
+    else
+      15.minutes.to_i
```

Comment:
We'll need #27948 (or something like it) for this to actually make sense, as an expiring access token without a way to refresh doesn't make much sense.

Reply:
I think this is a good idea - and see you have various PRs doing prep work ahead of it ... is there team buy-in on actually going this (expiring tokens) direction?

Reply:
My understand is "yes" for Mastodon 5.0.0 these should be default, but we'll need to support both for a period of time, hence needing to apply expiry conditionally.

So for example, if your client is using the `offline_access` scope (new) and it's a confidential client, then your access tokens will have an expiry, and refresh tokens will be used. If you're a public client (new), regardless of whether or not you request `offline_access`, your access token will have an expiry and support refresh tokens.